swi-prolog: release 8.1.0 for termux official repo

erlanger opened the pull request:
Hi @Grimler91,

   I appreciate your help! Here is the PR for SWI-Prolog ~7.7.23~ as we discussed.

~NOTE: when 7.7.24 is released the `calculate_alignment.patch` will not be needed.~

**Update**: 7.7.24 was released and this PR is up to date with that release.
**Update 2**: 7.7.25 was released and this PR is up to date with that release.

erlanger commented:
Seems cacert was updated, travis says:
```
Wrong checksum for https://curl.haxx.se/ca/cacert.pem:
Expected: 86695b1be9225c3cf882d283f05c944e3aabbc1df6428a4424269a93e997dc65
Actual:   4d89992b90f3e177ab1d895c00e8cded6c9009bec9d56981ff4f0a59e9cc56d6
```

erlanger commented:
The ca-certificates problem will  be solved by merging PR #3118 from @xeffyr

erlanger commented:
Closing and re-opening PR to trigger Travis re-build.

Grimler91 commented:
I think you need to add a commit to trigger a re-build in travis. You can rebase against master to trigger a rebuild (and fix ca-certificates error)

erlanger commented:
Thanks, closing the PR and reopening worked fine. Travis passed. What would be the next step?

erlanger commented on packages/ossp-uuid/uuid-aarch64.patch:
```diff
@@ -0,0 +1,1588 @@
+ #! /bin/sh
+ # Attempt to guess a canonical system name.
+ #   Copyright (C) 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999,
+-#   2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008
```
That patch comes directly from [Redhat's fedora](https://src.fedoraproject.org/cgit/rpms/uuid.git/plain/uuid-aarch64.patch) and I would rather not change it, to simplify maintenance.

Grimler91 replied:
Great. 

> and I would rather not change it

Yeah no problem, just wanted to know the origin/status of it

erlanger commented on packages/swi-prolog/build.sh:
```diff
@@ -0,0 +1,85 @@
+# this build for the artifacts needed to build the
+# Android version
+termux_step_host_build () {
+        termux_setup_ninja
```
Sure, I'll fix this.

erlanger commented on packages/ossp-uuid/build.sh:
```diff
@@ -0,0 +1,9 @@
+TERMUX_PKG_SHA256=11a615225baa5f8bb686824423f50e4427acd3f70d394765bdff32801f0fd5b0
+TERMUX_PKG_BUILD_IN_SRC=yes
+termux_step_pre_configure() {
+   export ac_cv_va_copy=C99
```
Sure, I'll change it to tabs.

erlanger commented on packages/swi-prolog/unstable-build.sh:
```diff
@@ -0,0 +1,170 @@
+
```
Sure, I'll delete it.

erlanger commented on packages/swi-prolog/build.sh:
```diff
@@ -0,0 +1,85 @@
+        termux_setup_ninja
+        termux_setup_cmake
+
+        sudo dpkg --add-architecture i386
```
~I think i686 is enough. I'll fix this.~
**Update**: i686 doesn't work because the zlib1g-dev:i686 pkg does not exist in ubuntu.

What do you think about doing this in two stages to keep things clean:

1. Keep the `dpkg --add-architecture` in the SWI-Prolog PR for now, until we get the SWI-Prolog merged.
2. On a separate PR provide a patch for `setup-ubuntu.sh` once SWI-Prolog is merged. 

The reason is that I think `setup-ubuntu.sh` has to do with all packages  (not just swi-prolog), so I would rather keep a clear separation of concerns to guarantee better stability for termux and easier maintenance. Let me know what you think.

Grimler91 replied:
> because the zlib1g-dev:i686 pkg does not exist in ubuntu.

Oops, my mistake. 

Host-build fails for me on arm with 
```
/usr/include/openssl/hmac.h:13:11: fatal error: openssl/opensslconf.h: No such file or directory
```
So I guess libssl-dev:i386 is needed as well

Grimler91 replied:
Fixing that gives 
```
FAILED: home/boot32.prc 
cd /home/builder/.termux-build/swi-prolog/build/src && /home/builder/.termux-build/_cache/cmake-3.13.1/bin/cmake -E remove -f /home/builder/.termux-build/swi-prolog/build/home/boot32.prc && /home/builder/.termux-build/swi-prolog/build/src/swipl -q -O -o /home/builder/.termux-build/swi-prolog/build/home/boot32.prc -b /home/builder/.termux-build/swi-prolog/build/home/boot/init.pl
/bin/sh: 1: /home/builder/.termux-build/swi-prolog/build/src/swipl: Exec format error
```
The host tries to run a binary built for arm (weird that aarch64 works though, should be the same problem there(?)).

erlanger replied:
It acually also needs openssl installed on the host (it needs to run the openssl binary to produce some test files). But if it doesn't find it, it will not build the test files.

The error you are getting about `exec format error` happens when you have an old `host-build` directory with the  wrong word size (e.g. the host-build directory was 64bit when building for arm, which is 32bit). So the swi-prolog tries to use the current build (which won't run on the host, because it is an `arm` binary. This is solved by removing the `host-build` directory.

For this reason we remove the `host-build` directory after the install, but since the build had failed for you, the `host-build` directory was not removed automatically.

I'll push a patch with openssl and libssl-dev:i386

erlanger commented:
Okay, pushed patches with the changes discussed above. Let me know what you think.

Grimler91 commented on packages/swi-prolog/build.sh:
```diff
@@ -0,0 +1,85 @@
+termux_step_post_make_install () {
+	mkdir -p ${_TMP_DIR}
+
+	# Remove host build because future buiilds may be
```
Tiny typo, buiilds

erlanger replied:
Right! You have good eyes!

Grimler91 commented on packages/swi-prolog/build.sh:
```diff
@@ -0,0 +1,85 @@
+	   -G "Ninja"					     \
+	   -DINSTALL_DOCUMENTATION=OFF			     \
+	   -DSWIPL_PACKAGES=ON				     \
+	   -DGMP=OFF					     \
```
I guess it should be `USE_GMP=off` here?

erlanger replied:
You're good!

erlanger commented:
Okay, fixes pushed.

erlanger commented:
Okay, pushed fixes adding `openssl` and `libssl-dev:i386` I guess  i had added them manually a while ago.

erlanger commented:
Okay, I just ran a `./build-package.sh -f -a all swi-prolog` with a clean container and it worked fine. Let me know if everything works fine for you also.

BTW, I really appreciate your help, but most of all those very sharp eyes you have :)

Grimler91 commented:
@erlanger interesting, I'm still getting errors, for arm and i686, but now related to ssl version I guess.
```
In file included from /home/builder/.termux-build/swi-prolog/src/packages/ssl/crypto4pl.c:45:
/home/builder/.termux-build/swi-prolog/src/packages/ssl/cryptolib.c:52:1: error: static declaration of ‘EVP_MD_CTX_free’ follows non-static declaratio
n
 EVP_MD_CTX_free(EVP_MD_CTX *ctx)
 ^~~~~~~~~~~~~~~
In file included from /usr/include/openssl/hmac.h:15,
                 from /home/builder/.termux-build/swi-prolog/src/packages/ssl/crypto4pl.c:40:
/usr/include/openssl/evp.h:536:6: note: previous declaration of ‘EVP_MD_CTX_free’ was here
 void EVP_MD_CTX_free(EVP_MD_CTX *ctx);
```

I will try deleting and recreating the docker image

erlanger commented:
> @erlanger interesting, I'm still getting errors, for arm and i686, but now related to ssl version I guess.
> 
> ```
> In file included from /home/builder/.termux-build/swi-prolog/src/packages/ssl/crypto4pl.c:45:
> /home/builder/.termux-build/swi-prolog/src/packages/ssl/cryptolib.c:52:1: error: static declaration of ‘EVP_MD_CTX_free’ follows non-static declaratio
> n
>  EVP_MD_CTX_free(EVP_MD_CTX *ctx)
>  ^~~~~~~~~~~~~~~

that is very strange... was that with the last patch I pushed?

erlanger commented:
To aid in debugging, the output of an `arm` build in a clean container is [here](https://gist.github.com/erlanger/5abfa808e81de909588a6493b27221c0).

You'll notice the first line says:
```
Creating new container...
```

and at the end
```
termux-elf-cleaner: Removing the DT_VERDEFNUM dynamic section entry from './lib/swipl/lib/arm-linux/ssl4pl.so'
termux-elf-cleaner: Removing the DT_VERDEF dynamic section entry from './lib/swipl/lib/arm-linux/ssl4pl.so'
termux - build of 'swi-prolog' done
```

We can `diff` your build output against that one to try to debug the problem.

erlanger commented:
Hmm..I think my branch is rebased properly, it is based on commit `b79e97a8` which is the latest from `master`. Should I use something different from `master`?

```
$ git log -n 7 --oneline   
903f172b (HEAD -> termux-official) swi-prolog: release 7.7.23 for termux official
repo
b79e97a8 (upstream/master, master) msmtp: Update from 1.8.0 to 1.8.1
daa1a995 libgpg-error: Update from 1.32 to 1.33
e83cc70a doxygen: Use fossies for tarball
f281e0bc zsh: Add mapfile module (fixes #3116)
e75b2134 fakeroot: Formatting
b86d08b4 rust: Update from 1.30.1 to 1.31.0
```

---------
Your logs show something amiss about OpenSSL:
```
$ grep OpenSSL /tmp/swi-prolog-arm.log
Configuring OpenSSL version 1.1.1a (0x1010101fL) for android-arm
***   OpenSSL has been successfully configured                     ***
***   (If you are new to OpenSSL, you might want to consult the    ***
-- Found OpenSSL: /usr/lib/x86_64-linux-gnu/libcrypto.so (found version "1.1.1")
-- Looking for OpenSSL_version
-- Looking for OpenSSL_version - not found
```

In your log the last line says _"OpenSSL_version - not found"_, whereas look my log says _"OpenSSL_version - found"_:
```
$ grep OpenSSL /tmp/arm-out.txt  
Configuring OpenSSL version 1.1.1a (0x1010101fL) for android-arm
***   OpenSSL has been successfully configured                     ***
***   (If you are new to OpenSSL, you might want to consult the    ***
-- Found OpenSSL: /usr/lib/i386-linux-gnu/libcrypto.so (found version "1.1.0g")
-- Looking for OpenSSL_version
-- Looking for OpenSSL_version - found
-- Found OpenSSL: /data/data/com.termux/files/usr/lib/libcrypto.so (found version
"1.1.1a")
-- Looking for OpenSSL_version
-- Looking for OpenSSL_version - found
```

Also, when building for arm, it should use the i386 version of openssl. Your logs show:
```
-- Found OpenSSL: /usr/lib/x86_64-linux-gnu/libcrypto.so (found version "1.1.1")
```

My log shows:
```
-- Found OpenSSL: /usr/lib/i386-linux-gnu/libcrypto.so (found version "1.1.0g")
```

Notice my log says `i386` which is the proper thing, because the host build should be using the 32bit version for arm. This leads me to think that you have an old `host-build` directory around.

Things to try:

Did you delete the container? `docker container stop <sha> && docker container rm <sha>`
Also make sure you do `rm -rf ~/.termux-packages/swi-prolog/host-build` to make sure you have a clean slate (I typed this by hand so check it's the directory beforehand).

There is some mystery we stumbled upon :)

xeffyr commented:
> This leads me to think that you have an old host-build directory around.

I'm not sure about this.

> Things to try:
Did you delete the container? docker container stop <sha> && docker container rm <sha>
Also make sure you do rm -rf ~/.termux-packages/swi-prolog/host-build in the container to make sure you have a clean slate (I typed this by hand so check it's the directory beforehand).

I just removed image termux/package-builder with all associated containers and then run `scripts/run-docker.sh` and `./build-package.sh -a arm -f swi-prolog`.

Got this:
```
FAILED: packages/ssl/CMakeFiles/plugin_crypto4pl.dir/crypto4pl.c.o 
/usr/bin/cc -Dplugin_crypto4pl_EXPORTS -Ipackages/ssl -I/home/builder/.termux-build/swi-prolog/src/src/os -I/home/builder/.termux-build/swi-prolog/src/src -m32 -O2 -fPIC   -Wall -D__SWI_PROLOG__ -DSERVER_CERT_REQUIRED=TRUE -DCLIENT_CERT_REQUIRED=TRUE -MD -MT packages/ssl/CMakeFiles/plugin_crypto4pl.dir/crypto4pl.c.o -MF packages/ssl/CMakeFiles/plugin_crypto4pl.dir/crypto4pl.c.o.d -o packages/ssl/CMakeFiles/plugin_crypto4pl.dir/crypto4pl.c.o   -c /home/builder/.termux-build/swi-prolog/src/packages/ssl/crypto4pl.c
In file included from /home/builder/.termux-build/swi-prolog/src/packages/ssl/crypto4pl.c:45:
/home/builder/.termux-build/swi-prolog/src/packages/ssl/cryptolib.c:52:1: error: static declaration of ‘EVP_MD_CTX_free’ follows non-static declaration
 EVP_MD_CTX_free(EVP_MD_CTX *ctx)
 ^~~~~~~~~~~~~~~
In file included from /usr/include/openssl/hmac.h:15,
                 from /home/builder/.termux-build/swi-prolog/src/packages/ssl/crypto4pl.c:40:
/usr/include/openssl/evp.h:536:6: note: previous declaration of ‘EVP_MD_CTX_free’ was here
 void EVP_MD_CTX_free(EVP_MD_CTX *ctx);
```

libssl is installed:
```
ii  libssl-dev:amd64             1.1.1-1ubuntu2.1                  amd64        Secure Sockets Layer toolkit - development files
ii  libssl-dev:i386              1.1.1-1ubuntu2.1                  i386         Secure Sockets Layer toolkit - development files
ii  libssl1.1:amd64              1.1.1-1ubuntu2.1                  amd64        Secure Sockets Layer toolkit - shared libraries
ii  libssl1.1:i386               1.1.1-1ubuntu2.1                  i386         Secure Sockets Layer toolkit - shared libraries
```
But it's version is 1.1.1 which is incompatible with 1.1.0.

xeffyr commented:
@erlanger Do you use the latest version of docker image ? The latest one has Ubuntu 18.10 and libssl-dev v1.1.1.

erlanger commented:
> @erlanger Do you use the latest version of docker image ? The latest one has Ubuntu 18.10 and libssl-dev v1.1.1.

Ahh...I think you might found the culprit.... let me test

erlanger commented:
Okay, took me a while to track this down.

In the new Ubuntu,  CMake has a problem finding the 32 bit version of `libcrypto.so` but it finds the ssl headers. This causes SWI-Prolog to try to define some functions on its own, but the ssl headers are also included. This  is why we get the `static declaration follows non-static declaration` error: because SWI-Prolog function definitions are clashing with the ones in the headers.

In any case, I applied a patch that solves the problem. @Grimler91, @xeffyr  let me know if you are able to build now. I was able to build for `arm` and `aarch64` without problems.

Phew! This was a hard one to track down! Thank you guys for your tremendous help. I won't sing victory yet, until you are able to replicate the build :smirk:

erlanger commented:
## Update
I just ran `./build-package.sh -f -a all swi-prolog` successfully :smile:

xeffyr commented:
@erlanger Yes, now it builds successfully.

erlanger commented:
Great! Thanks! What is the next step?

erlanger commented:
Seems there was a temporary problem downloading `libxml2`:
```
Warning: Transient problem: FTP error Will retry in 2 seconds. 1 retries left.


  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0 5340k    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0

curl: (19) RETR response: 425

Download of ftp://xmlsoft.org/libxml2/libxml2-2.9.8.tar.gz failed (attempt 6/6)

ERROR: Failed to download ftp://xmlsoft.org/libxml2/libxml2-2.9.8.tar.gz
```

Closing and re-opening PR to trigger travis rebuild.

xeffyr commented:
Usually, there no need to make Travis rebuild. It is used only for development purposes and not for actual package build.

erlanger commented:
Thanks Xeffyr! I wanted to make sure it was building okay. 
Travis passed.

xeffyr commented:
Now seems okay. However, sudo-related lines like those
```
sudo dpkg --add-architecture i386
sudo apt-get -y update
sudo apt-get -y install openssl libssl-dev:i386 zlib1g-dev:i386
```
needs to be placed to [setup-ubuntu.sh](https://github.com/termux/termux-packages/blob/master/scripts/setup-ubuntu.sh).

 I guess @fornwall will review/merge it soon.

erlanger commented:
> Now seems okay. However, sudo-related lines like those
> 
> ```
> sudo dpkg --add-architecture i386
> sudo apt-get -y update
> sudo apt-get -y install openssl libssl-dev:i386 zlib1g-dev:i386
> ```
> 
> needs to be placed to [setup-ubuntu.sh](https://github.com/termux/termux-packages/blob/master/scripts/setup-ubuntu.sh).
> 

Yes, Grimler91 mentioned that, I wanted to do that change in a separate PR after SWI-Prolog is merged. 

The reason is that `setup-ubuntu.sh`  affects all packages, not just SWI-Prolog, and I wanted to keep a clear separation of concerns. It is also needed mostly for the 32bit host build, so it may need a little bit more thought than just putting them on `setup-ubuntu.sh`. 

Would that way of proceeding be OK with you?
 
> I guess @fornwall will review/merge it soon.

Great!

xeffyr commented:
> it may need a little bit more thought than just putting them on setup-ubuntu.sh.

This script is used when generating Docker image. Some users don't use docker image or even use an operating system where package manager is different (e.g. Arch, [setup-arch-linux.sh](https://github.com/termux/termux-packages/blob/master/scripts/setup-archlinux.sh)). So necessary packages should be definetly installed when creating a build environment and not in runtime.

erlanger commented:
> > it may need a little bit more thought than just putting them on setup-ubuntu.sh.
> 
> ...So necessary packages should be definetly installed when creating a build environment and not in runtime.

I see...yeah..I agree.  Would it be okay with you if we do the changes to `setup-ubuntu.sh` in a separate PR after SWI-Prolog is merged?

BTW, how is the docker image recreated? does it happen automatically on a merge that modifies `setup-ubuntu.sh`?

erlanger commented:
Hi @xeffyr, @Grimler91 ,
  Just pushed some updates:
* Moved the `sudo` lines to `setup-ubuntu.sh` as you requested
* Updated to version 7.7.25 just released

Let me know if it builds OK for you with the new docker image.

erlanger commented:
Rebased to make just one commit.

erlanger commented:
Hi @fornwall,

   Thank you very much for the work you have done with termux, it has benefited many people. 

Is there anything else that needs to be done for SWI-Prolog to be merged?

Thanks! Appreciate your help!

erlanger commented:
Thanks @its-pointless. I added the changes.

dram commented:
Any news about this? BTW, version 8.1.0 was released.

erlanger commented:
I'm waiting on fornwall to merge. Hopefully it will happen soon.

erlanger commented:
Thanks Xeffyr, could I ask why did it go into unstable packages? 100% of the SWI-Prolog tests pass. Thanks!

xeffyr commented:
For now, repository "unstable-packages" is not only for unstable but for **new** packages too.
Regarding when it will be in main repo ask in https://gitter.im/termux/dev. Since there are planned "global" changes + Fornwall is busy, no one knows when actually something gets merged.

erlanger commented:
Ahh I see, thanks!

jean commented:
So PR [was merged somehow](https://github.com/termux/termux-packages/pull/3129#issuecomment-459363484), but it's still open, with Travis failing? Seems like it's in the twilight zone ..

its-pointless commented:
Its in unstable-packages
